### PR TITLE
[IMPROVE] Improvements and fixes to stream source consumer creation on leadership change or on stream config source updates

### PR DESCRIPTION
**Pull request by jnmoyne**
- [X] Tests added
 - [X] Branch rebased on top of current main (`git pull --rebase origin main`)
 - [X] Changes squashed to a single commit (described [here](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
 - [x] Build is green in Travis CI
 - [X] You have certified that the contribution is your original work and that you license the work to the project under the [Apache 2 license](https://github.com/nats-io/nats-server/blob/main/LICENSE)

### CODE REVIEW

File: server/stream.go [hunk 1]
```diff
@@ -605,15 +605,32 @@ func (a *Account) addStreamWithAssignment(config *StreamConfig, fsConfig *FileSt
...
 	if ssi.External != nil {
-		ssi.iname = strconv.Itoa(i) + ":" + ssi.Name + ":" + getHash(ssi.External.ApiPrefix)
-	} else {
-		ssi.iname = strconv.Itoa(i) + ":" + ssi.Name
+		iName = iName + ":" + getHash(ssi.External.ApiPrefix)
+	}
+
+	filter := ssi.FilterSubject
+	// normalize filter and destination in case they are empty
+	if filter == "" {
```
Comment (derekcollison) [inline server/stream.go:619]: _EMPTY_ vs "" here.

File: server/stream.go [hunk 2]
```diff
@@ -605,15 +605,32 @@ func (a *Account) addStreamWithAssignment(config *StreamConfig, fsConfig *FileSt
...
+		filter = ">"
```
Comment (derekcollison) [inline server/stream.go:620]: Use const, fwcs

File: server/stream.go [hunk 3]
```diff
@@ -605,15 +605,32 @@ func (a *Account) addStreamWithAssignment(config *StreamConfig, fsConfig *FileSt
...
+	}
+	destination := ssi.SubjectTransformDest
+	if destination == "" {
```
Comment (derekcollison) [inline server/stream.go:623]: Same here.

File: server/stream.go [hunk 4]
```diff
@@ -605,15 +605,32 @@ func (a *Account) addStreamWithAssignment(config *StreamConfig, fsConfig *FileSt
...
+		destination = ">"
 	}
+
+	iName = iName + " " + filter + " " + destination
```
Comment (derekcollison) [inline server/stream.go:627]: strings.Join()?

File: server/stream.go [hunk 5]
```diff
@@ -2890,27 +2882,38 @@ func streamAndSeqFromAckReply(reply string) (string, uint64) {
...
+	nFields := len(fields)
+
+	if nFields != 2 && nFields <= 3 {
+		return _EMPTY_, _EMPTY_, 0
 	}
-	return fields[0], uint64(parseAckReplyNum(fields[1]))
+
+	sName := fields[0]
+	if nFields >= 4 {
+		iName := sName + " " + fields[2] + " " + fields[3]
```
Comment (derekcollison) [inline server/stream.go:2906]: Same?

File: server/stream.go (hunk 1 above)
Comment (jnmoyne) [reply server/stream.go]: Done, sorry I forgot to check for _EMPTY_ and fwcs, it's not a reflex for me yet :)

File: server/stream.go (hunk 3 above)
Comment (jnmoyne) [reply server/stream.go]: Done

File: server/stream.go (hunk 4 above)
Comment (jnmoyne) [reply server/stream.go]: Ok, done.

File: server/stream.go (hunk 2 above)
Comment (jnmoyne) [reply server/stream.go]: Done

File: server/stream.go (hunk 5 above)
Comment (jnmoyne) [reply server/stream.go]: Done

**Comment by derekcollison** [comment]
This ready to review?

**Comment by jnmoyne** [comment]
Apologies, yes it is!

**Review by derekcollison — APPROVED** [review]
LGTM

**Pull request merged**